library/Ivoz/Provider/Domain/Service/FeaturesRelBrand/AvoidFeaturesInCompanyRemoval.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Ivoz\Provider\Domain\Service\FeaturesRelBrand;
+
+use Ivoz\Provider\Domain\Model\Company\CompanyInterface;
+use Ivoz\Provider\Domain\Model\Company\CompanyRepository;
+use Ivoz\Provider\Domain\Model\FeaturesRelBrand\FeaturesRelBrandInterface;
+
+class AvoidFeaturesInCompanyRemoval implements FeaturesRelBrandLifecycleEventHandlerInterface
+{
+    public const PRE_REMOVE_PRIORITY = self::PRIORITY_HIGH;
+
+    public function __construct(
+        private CompanyRepository $companyRepository,
+    ) {
+    }
+
+    /** @return array<array-key, int> */
+    public static function getSubscribedEvents()
+    {
+        return [
+            self::EVENT_PRE_REMOVE => self::PRE_REMOVE_PRIORITY
+        ];
+    }
+
+    /**
+     * @throws \DomainException
+     */
+    public function execute(FeaturesRelBrandInterface $relBrand): void
+    {
+        $brand = $relBrand->getBrand();
+        $feature = $relBrand->getFeature();
+        if (!$brand) {
+            return;
+        }
+
+        /** @var CompanyInterface[] $companies */
+        $companies = $this->companyRepository
+            ->findByBrandId(
+                (int) $brand->getId()
+            );
+
+        foreach ($companies as $company) {
+            $companyFeatures = $company->getRelFeatures();
+            foreach ($companyFeatures as $companyRelFeature) {
+                $featureRelCompanyId = $companyRelFeature
+                    ->getFeature()
+                    ->getId();
+                if ($feature->getId() === $featureRelCompanyId) {
+                    throw new \DomainException('Cannot remove a feature used by an associated company.');
+                }
+            }
```

Comment:
This can be done in one single repository query. Something like
`SELECT x FROM Companies INNER JOIN CompanyRelFeatures where brandId = y`